test/unit/gateways/mit_test.rb

```diff
@@ -230,14 +230,16 @@ def post_scrubbed
     -> "Server: \r\n"
     -> "Set-Cookie: UqZBpD3n=v1I4cyJQ__N2M; Expires=Mon, 06-Sep-2021 19:03:38 GMT; Path=/; Secure; HttpOnly\r\n"
     -> "\r\n"
-    response: {"folio_cdp":"095492846","auth":"928468","response":"approved","message":"0C- Pago aprobado (test)","id_comercio":"147","reference":"721","amount":"11.15","time":"19:02:08 06:09:2021","operation":"Authorize"}read 320 bytes
+    reading 320 bytes...
```

Comment:
We left that portion of the response encrypted due to an issue in which the encrypted responses from the gateway were altered (split into multiple sub strings) by the transcription and were left unable to be decrypted. This happens very rarely but it is the reason some of your remote tests had issues. Is there a way we can identify under what circumstances the generated transcription splits a string into substrings?

An example of what we saw would be like this

A response string would typically look like this:

-> "BnuAgMOx9USBreICk027VY2ZqJA7xQcRT9Ytz1WpabDnqIglj43J/I03pKLtDlFrerKIAzhW1YCroDOS7mvtA5YnWezLstoOK0LbIcYqLzj1dCFW2zLb9ssTCxJb7ZmEQdzQdl8pyY4qC0QQ0JrOrsSA9QfX1XhkdcSVnttQV1cEooL8/6EsVFCb6yVIMhVnGL6GRCc2J+rPigHsljLWRovgRKqFIURJjNWbbbbbDRPG2hCNKsabM/lE2DFtKLMs4J5iwY9HiRbrAMG6BaGNiQ=="

But the generated transcription we receive would look like this:

-> "BnuAgMOx9USBreICk027VY2ZqJA7xQcRT9Ytz1WpabDnqIglj43J/I03pKLtDlFrerKIAzhW1YCroDOS7mvtA5YnWezLstoOK0LbIcYqLzj1dCFW2zLb9ssTCxJb7ZmEQdzQdl8pyY4qC0QQ0JrOrsSA9QfX"
-> "1XhkdcSVnttQV1cEooL8/6EsVFCb6yVIMhVnGL6GRCc2J+rPigHsljLWRovgRKqFIURJjNWbbbbbDRPG2hCNKsabM/lE2DFtKLMs4J5iwY9HiRbrAMG6BaGNiQ=="

We don't know why sometimes we received a split encrypted answer and when we do not, even if the transaction is identical. If we decrypt the responses again this specific portion of the transcription would generate issues in the scrubbing function. We have passed this to the gateway in case it is an issue of theirs, but I would like to know if there's a case in which the transcription generation would split a string like that.

Reply:
I just sent this message to Gabriela in an email but in case it doesn't get forwarded quickly:

We looked up the errors in the stack trace from the transaction you linked in a support thread:
Full error: "full_error_with_trace"=>"#<OpenSSL::Cipher::CipherError: wrong final block length>\n

The input byte length is not actually a multiple of 16 like the cipher expects. This is supported by running the remote test test_transcript_scrubbing and the test fails with ArgumentError: iv must be 16 bytes. You can run these remote tests and see this. 
Stacktrace:
ArgumentError: iv must be 16 bytes /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `iv=' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `decrypt' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:190:in `block in scrub' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `loop' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `scrub' /Users/bsmith/dev/active_merchant/test/remote/gateways/remote_mit_test.rb:123:in `test_transcript_scrubbing'


The response splitting is not something that happens in ActiveMerchant, this is happening from the gateway before it gets to ActiveMerchant. A suggestion is to inspect where the encrypted response is concatenated and converted to base64 (here: https://github.com/activemerchant/active_merchant/blob/00f1189edab256b7a7d898368c0eeece2ddcbb1b/lib/active_merchant/billing/gateways/mit.rb#L75). 
 
After speaking with a colleague who is more versed in security and ciphers, they gave the following suggestions:

Possible solutions:
Adjust padding using the PKCS7 padding scheme to make up for byte lengths that are lesser than 16, with something like: engine.padding = 1 # Sets PKCS7 padding (src: [So What Is PKCS#7?](https://medium.com/asecuritysite-when-bob-met-alice/so-what-is-pkcs-7-daf8f4423fd1) )
Update the engine from CBC to prevent the need for padding in the first place (src: [AES-256, CBC, plaintext length is multiple of block size: Does PKCS #7 padding weaken the encryption?](https://crypto.stackexchange.com/a/82002) , [Incorrect AES Implementation Leaves System Vulnerable](https://www.winmill.com/blog/2022/08/05/incorrect-aes-implementation-leaves-system-vulnerable/) ).
:note: The engine from above is defined as engine = OpenSSL::Cipher::AES128.new(:CBC). I also wonder if they should graduate from AES-128 to AES-256.


I will be out of office next week but my teammates will be keeping an eye on this PR.

Reply:
Please see the comments above.

Reply:
Hi @rachelkirk I think that we had found what it is the problem and we are looking for ways to solve it.

When the tests are ran the transcript is being stored on the transcript variable:

https://github.com/activemerchant/active_merchant/blob/92046f1680094d1b70557040d1c1edb193f7948d/test/remote/gateways/remote_mit_test.rb#L119

```
def test_transcript_scrubbing
    transcript = capture_transcript(@gateway) do
      @gateway.purchase(@amount, @credit_card, @options_success)
    end

    clean_transcript = @gateway.scrub(transcript)
    assert_scrubbed(@credit_card.number, clean_transcript)
    assert_scrubbed(@credit_card.verification_value, clean_transcript)
    assert_scrubbed(@gateway.options[:api_key], clean_transcript)
    assert_scrubbed(@gateway.options[:key_session], clean_transcript)
end
```

The method called `capture_transcript`, captures all the communication between Spreedly and MIT and stores it on the variable transcript (keep in mind that the method capture_transcript is not part of MIT integration)

Then on `@gateway.scrub` the variable `transcript` is passed. 

When reading the content of the variable `transcript` there are requests that sometimes are serialized on the `transcript` as with the full body of the response and sometimes the body of the response is being "broken". For example:

*Full*
```
-> "\r\n"
reading 344 bytes...
-> "VW0iCiVNSsdkxn0rBnESgK8lLiLnGxH/csfK4pnDVssQ1+QEs/XHWJMn9IMxnJehm1rn2UnOpq8Q4CAWrh9/xcVYmjc0N4BTrAISGcLHXQHskNth4cf3woREZ2lD2pOok2gmCbC/aEhhsHqO21G5NsZj09+5p7HiCTutiFnzFebkrBqJEV+g1Vm/oJHqMweU7wewFY6w4vHfKnSCBbTXfHygvbT98gugfm2/H6Y+d9EEOsu0nJb9eh9y9PC4tweoFA3vVe0spl0WIg14vexITkVN1HZIA5onw5SZtqW76JktGfRd7+K7L1oBjhWMjdmw2zYOpW4rEnQoAuzG90/sBw=="
read 344 bytes
Conn close
```


*Broken*
```
reading 344 bytes...
-> "lPRvJkNr1bg1"
-> "kHdqguEDHKCws4iDEEtLGaOz2/FZe7SOD4RliKkJzz61Lye80ts5ro13PKke+gEuHcR3hAu66m/bJRQCggmHFt3nIPyVrj9elZIXwaAcA1aSoijwTSzNwjjVQg+RQL9gYBweb86QvPUkmsKQFpyAP9lGQ1TTJLVKYGXqq+7cLiAaxdhehlx1N2Knckg7BDM8gP0j9ExsSXopaiBQ9ZZTzkIv2UURoFc+4AMMco/cFFXf9RyHkzER21EtxIs600ENcwhgsJ8GRjAhx6nNz+c5R5nCXK0xcsR/+KDpIHZUQEulrrb0VeLiJElsvhc++Uk5EHZp4Yq9mQ=="
read 344 bytes
```

Notice that on the "Broken" it shows two `->`.

So this is the reason why when the tests are running `clean_transcript = @gateway.scrub(transcript)` it fails. Because when integration is reading the response from the MIT server from the `transcript` it fails  since its not the response from the MIT Server: https://github.com/activemerchant/active_merchant/blob/92046f1680094d1b70557040d1c1edb193f7948d/lib/active_merchant/billing/gateways/mit.rb#L173

Do you have any recommendation regarding on how to avoid that the response body is added incorrectly into the transcript?

Thanks!

Reply:
I'm still looking into this, but I have not seen any instances of ActiveMerchant splitting a response like this before. This may have something to do with the way the encrypt and decrypt methods are written. There are only a few gateways out of 150+ in ActiveMerchant that have encrypt and decrypt methods defined like this. The helper method that you wrote is returning the auth and capture response simultaneously for a purchase since it's a `MultiResponse.run` of auth and capture for purchase. I'm doing more testing locally to see if that will be an issue.